File: 740-delete-and-earn/740-delete-and-earn.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int n = 10002;
        vector<int> dp(n, 0);
        vector<int> freq(n, 0);
        for(int i=0;i<nums.size();i++) {
            freq[nums[i]]++;
        }
        dp[0] = 0;
        dp[1] = 1 * freq[1];
        
        for(int i=2;i<10002;i++) {
            dp[i] = max(dp[i-1], dp[i-2] + i*freq[i]);
        }
        return dp[n-1];
    }
};

File: 740-delete-and-earn/740-delete-and-earn.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        int[] freq = new int[10002];
        for(int i=0;i<nums.length;i++) {
            freq[nums[i]]++;
        }
        HashMap<Integer, Integer> memo = new HashMap<Integer, Integer>();
        return maxPoints(0, freq, memo); // 0-> cur idx
    } 
    private int maxPoints(int curIdx, int[] freq, HashMap<Integer, Integer> memo) {
        if(curIdx >= freq.length) {
            return 0;
        }
        int curKey = curIdx;
        if(memo.containsKey(curKey))
            return memo.get(curKey);
        
        int del = curIdx*freq[curIdx] + maxPoints(curIdx+2, freq, memo);
        int nodel = maxPoints(curIdx+1, freq, memo);
        
        memo.put(curKey, Math.max(del, nodel));
        return memo.get(curKey);
    }
}